### PR TITLE
minor refactorings around `InferCtxt::enter`

### PR DESCRIPTION
**Pull request by nikomatsakis**
Just a few small simplifications.

r? @eddyb

### CODE REVIEW

File: src/librustc_typeck/check/mod.rs [hunk 1]
```diff
@@ -189,7 +189,7 @@ impl<'a, 'tcx> MaybeInProgressTables<'a, 'tcx> {
 /// `bar()` will each have their own `FnCtxt`, but they will
 /// share the inherited fields.
 pub struct Inherited<'a, 'gcx: 'a+'tcx, 'tcx: 'a> {
-    infcx: InferCtxt<'a, 'gcx, 'tcx>,
+    infcx: &'a InferCtxt<'a, 'gcx, 'tcx>,
```
Comment (eddyb) [inline src/librustc_typeck/check/mod.rs]: Why the extra indirection?

File: src/librustc/infer/mod.rs [hunk 2]
```diff
@@ -404,43 +404,42 @@ impl fmt::Display for FixupError {
...
     pub fn with_fresh_in_progress_tables(mut self, table_owner: DefId) -> Self {
-        self.fresh_tables = Some(RefCell::new(ty::TypeckTables::empty(Some(table_owner))));
+        self.fresh_table_owner = Some(table_owner);
         self
     }
 
-    pub fn enter<F, R>(&'tcx mut self, f: F) -> R
-        where F: for<'b> FnOnce(InferCtxt<'b, 'gcx, 'tcx>) -> R
-    {
+    pub fn enter<'tcx, R>(&'tcx mut self, f: impl FnOnce(&InferCtxt<'_, 'gcx, 'tcx>) -> R) -> R {
```
Comment (eddyb) [inline src/librustc/infer/mod.rs]: I guess it started here - why?

Comment (nikomatsakis) [reply src/librustc/infer/mod.rs]: I can revert that part if you like -- I was actually working towards a goal with this patch series that didn't work out, so it's not really important either way (I just figured I'd land some of the incidental changes).

That said, I always get annoyed that enter gives me ownership of an `InferCtxt`, when almost everything wants a reference. I'd be quite surprised if the indrection around `Inherited` cost performance.

**Comment by eddyb** [comment]
r=me with travis failure fixed

**Comment by nikomatsakis** [comment]
Sigh. I should really rebase this.

**Comment by shepmaster** [comment]
Ping from triage, @nikomatsakis ! We agree you should rebase this ❤️. Might as well respond to the review feedback you got while you are at it!

**Comment by nikomatsakis** [comment]
Eh. I don't care enough to rebase this. =)

**Pull request closed without merging**